### PR TITLE
netty: create adaptive cumulator

### PR DESCRIPTION
**Pull request by sergiitk**
Creates "Adaptive" cumulator: cumulate ByteBuf's by dynamically switching between merge and compose strategies.

This cumulator applies a heuristic to make a decision whether to track a reference to the buffer with bytes received from the network stack in an array ("zero-copy"), or to merge into the last component (the tail) by performing a memory copy.

It is necessary as a protection from a potential attack on the COMPOSITE_CUMULATOR. Consider a pathological case when an attacker sends TCP packages containing a single byte of data, and forcing the cumulator to track each one in a separate buffer. In this case we'll be paying a memory overhead for each buffer, as well as extra compute to read the cumulation.

Implemented heuristic establishes a minimal threshold for the total size of the tail and incoming buffer, below which they are merged. The sum of the tail and the incoming buffer is used to avoid a case where attacker alternates the size of data packets to trick the cumulator into always selecting compose strategy.

Merging strategy attempts to minimize unnecessary memory writes. When possible, it expands the tail capacity and only copies the incoming buffer into available memory. Otherwise, when both tail and the buffer must be copied, the tail is reallocated (or fully replaced) with a new buffer of exponentially increasing capacity (bounded to minComposeSize) to ensure runtime O(n^2) amortized to O(n).

### CODE REVIEW

**Comment by ejona86** [comment]
FYI @Scottmitch @njhill . Note that this opens up opportunities for larger frame sizes. We would be interested to getting this into Netty, but we were going to be trying things out first.

**Comment by normanmaurer** [comment]
This sounds interesting... keep us posted

**Review by ejona86 — APPROVED** [review]

File: netty/src/test/java/io/grpc/netty/NettyTestUtil.java [hunk 1]
```diff
@@ -56,6 +57,13 @@ static ByteBuf compressionFrame(byte[] data) {
     return buf;
   }
 
+  // Buffer safe release: release until it's free.
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyTestUtil.java:60]: No, please no. This is asking to be misused, and I don't think it is appropriate even in the test. I do think we should clean up the buffers, but if the reference counting is wrong the test should fail.

File: netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java [hunk 2]
```diff
@@ -84,6 +86,14 @@ void addInput(ByteBufAllocator alloc, CompositeByteBuf composite, ByteBuf in) {
           composite.addFlattenedComponents(true, in);
         }
       };
+
+      // Throws an error on adding incoming buffer.
+      throwingCumulator = new NettyAdaptiveCumulator(0) {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java:91]: nit: normally this sort of thing we'd just do inline at the declaration and not in the `@Before`. This is fine though.

**Review by ejona86 — APPROVED** [review]

File: netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java (hunk 2 above)
Comment (sergiitk) [reply netty/src/test/java/io/grpc/netty/NettyAdaptiveCumulatorTest.java:91]: I'll get it fixed separately to not delay this.

**Pull request merged**